File: Python/command-finder.py
import os, re, time, json, pathlib, zipfile

# possible output types:
# plain
# json
output_type = 'plain'

path = pathlib.PurePath(os.getcwd())

if 'Add-Ons' in os.listdir():
    addon_path = path / 'Add-Ons'
else:
    addon_path = path

regex = re.compile(b'\s*function serverCmd([\w\d]+)\s*\(([%\w\d,\s]+)*\)', re.I)
results = {}


start = time.clock()

addon_list = [f for f in os.listdir(str(addon_path)) if f[-3:] == 'zip']

for zip_file_name in addon_list:
    zip = zipfile.ZipFile(str(addon_path / zip_file_name))
    file_list = zip.namelist()
    results[zip_file_name[:-4]] = {}
    for file_name in file_list:
        if file_name[-2:] == 'cs':
            file = zip.open(file_name)
            results[zip_file_name[:-4]][file_name[:-3]] = {}
            for line in file:
                match = regex.match(line)
                if match:
                    results[zip_file_name[:-4]][file_name[:-3]][match.group(1).decode()] = [m.strip() for m in match.group(2).decode().split(',')]
            if not results[zip_file_name[:-4]][file_name[:-3]]:
                del results[zip_file_name[:-4]][file_name[:-3]]
            file.close()
    if not results[zip_file_name[:-4]]:
        del results[zip_file_name[:-4]]
    zip.close()

end = time.clock()


if output_type == 'plain':
    output_file_name = 'server_commands.txt'
    server_cmd_pl = open(output_file_name, 'w+')
    for addon, script_list in sorted(results.items()):
        if server_cmd_pl.tell() != 0:
            server_cmd_pl.write('\n{0}\n'.format(addon))
        else:
            server_cmd_pl.write('{0}\n'.format(addon))
        for script in script_list.values():
            for command in sorted(script.keys()):
                server_cmd_pl.write('    {0}\n'.format(command))
    server_cmd_pl.close()
else:
    output_file_name = 'server_commands.json'
    serv_cmd_json = open(output_file_name, 'w+')
    json.dump(results, serv_cmd_json, sort_keys=True, indent=4)
    serv_cmd_json.close()

print('Results can be found in {0}.'.format(output_file_name))
print('Time taken: {0}'.format(end - start))
input('Press enter to exit.')

File: Java/CommandFinder.java
import java.util.*;
import java.util.regex.*;
import java.util.zip.*;
import java.io.*;
import java.nio.file.*;
import java.nio.file.attribute.*;

import com.google.gson.*;


public class CommandFinder {
	/* JSON should look roughly like this:
	 * 
	 * {
	 * 	 "Addon_Name": {
	 *     "filename": {
	 *       "cmdName": [
	 *         "argument1",
	 *         "argument2"
	 *       ]
	 *     }
	 *   }
	 * }
	 * 
	 */
	
	
	HashMap<String, HashMap<String, HashMap<String, ArrayList<String>>>> output = new HashMap<>();
	
	static String regexPattern= "\\s*function serverCmd([\\w\\d]+)\\s*\\(([%\\w\\d,\\s]+)*\\)";
	static Pattern regex = Pattern.compile(regexPattern, Pattern.CASE_INSENSITIVE);
	
	Path path;
	ArrayList<String> fileList = new ArrayList<String>();
	
	static Gson json = new GsonBuilder().setPrettyPrinting().create();
	
	
	public CommandFinder(String addonPath) {
		path = Paths.get(addonPath);
		
		try {
		    Files.walkFileTree(path, new SimpleFileVisitor<Path>() {
		    	@Override
		        public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) {
		    		if (file.toString().endsWith(".zip")) {
		    			fileList.add(file.toString());
		    		}		    		
		            return FileVisitResult.CONTINUE;
		        }
		    });
		}
		catch (IOException e) {
		    System.out.println("Something totally bad just happened: " + e.getMessage());
		}
	}
	
	public void searchFiles() {
		for (String fileName : fileList) {
			try {
				ZipFile zip = new ZipFile(fileName);
				Enumeration<? extends ZipEntry> entries = zip.entries();
				
				String addOnName = fileName.substring(fileName.lastIndexOf("\\") + 1);
				addOnName = addOnName.substring(0, addOnName.length() - 4);
				HashMap<String, HashMap<String, ArrayList<String>>> addOnMap = new HashMap<>();
				output.put(addOnName, addOnMap);
				
				while (entries.hasMoreElements()) {
					try {
						ZipEntry entry = entries.nextElement();
						String entryName = entry.getName();
						if (entryName.endsWith(".cs")) {
							InputStream zipIn = zip.getInputStream(entry);
							BufferedReader reader = new BufferedReader(new InputStreamReader(zipIn, "UTF-8"));
							
							HashMap<String, ArrayList<String>> fileMap = new HashMap<>();
							addOnMap.put(entryName.substring(0, entryName.length() - 3), fileMap);
							
							String line = null;
						    while ((line = reader.readLine()) != null) {
						    	ArrayList<String> argList = new ArrayList<>();
						    	Matcher matcher = regex.matcher(line);
						    	
						    	if (matcher.matches()) {
						    		String cmdName = matcher.group(1);
							    	fileMap.put(cmdName, argList);
							    	
							    	for (String argument : matcher.group(2).split(",")) {
							    		argList.add(argument.trim());
							    	}
						    	}
						    }
							
						    
							reader.close();
							zipIn.close();
						}
					}
					catch (ZipException e) {
						System.out.println("Error with zip entry: " + e.getMessage());
					}
				}
				
				zip.close();
			}
			catch (ZipException e) {
				System.out.format("Error opening zip file %s: %s",
						fileName.substring(fileName.lastIndexOf("\\") + 1),
						e.getMessage());
				System.out.println();
			}
			catch (IOException e) {
				System.out.format("Something bad happened: %s: %s",
						fileName.substring(fileName.lastIndexOf("\\") + 1),
						e.getMessage());
				System.out.println();
			}
		}
	}
	
	public void cleanOutput() {
		for (String addOnName : output.keySet()) {
			HashMap<String, HashMap<String, ArrayList<String>>> addOnMap = output.get(addOnName);
			
			Iterator<String> addOnIter = addOnMap.keySet().iterator();
			while (addOnIter.hasNext()) {
				String fileName = addOnIter.next();
				
				HashMap<String, ArrayList<String>> fileMap = addOnMap.get(fileName);
				
				if (fileMap.isEmpty())
					addOnIter.remove();
			}
		}
		
	}
	
	public void writeJson() {
		try {
			Writer writer = new FileWriter("server_commands.json");
			json.toJson(output, writer);
			writer.close();
		}
		catch (IOException e) {
			System.out.println("Couldn't write the JSON file, for some reason: " + e.getMessage());
			e.printStackTrace();
		}
	}
	
	public static void main(String[] args) {
		double originalTime = System.nanoTime() / 1000000000.0;
		
		CommandFinder cmds = new CommandFinder(System.getProperty("user.dir") + "/Add-Ons");
		cmds.searchFiles();
		cmds.cleanOutput();
		cmds.writeJson();
		
		double newTime = System.nanoTime() / 1000000000.0;
		double difference = newTime - originalTime;
		
		System.out.println("Results can be found in server_commands.json.");
		System.out.println("Time taken: " + difference);
		System.out.print("Press enter to exit.");
		try {
			System.in.read();
		}
		catch (IOException e) {
			System.out.println("For some reason, trying to read input did not work...");
			System.out.println(e.getMessage());
		}
	}
}
